returnn/tf/layers/basic.py

```diff
@@ -854,44 +858,87 @@ class SliceNdLayer(_ConcatInputLayer):
 
   def __init__(self, start, size, min_size=None, **kwargs):
     """
-    :param LayerBase start:
+    :param LayerBase start: (B,...)
     :param int|None size: if None, it uses the max possible size, and it becomes a dynamic axis
-    :param int|None min_size: if size is None, but we want to have a min-size, set this
+    :param int|None min_size: if size is None, but we want to have a min-size
     """
     super(SliceNdLayer, self).__init__(**kwargs)
-    from returnn.tf.util.basic import slice_nd, where_bc, expand_multiple_dims, DimensionTag
+    from returnn.tf.util.basic import where_bc, expand_multiple_dims
     x = self.input_data.copy_as_batch_major()
-    assert x.time_dim_axis == 1, "currently only time-axis==1 supported"
-    seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None
+    seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None  # (B,) or None
     self.start = start
-    assert start.output.have_batch_axis() and start.output.batch_shape == (None,)
-    start = start.output.get_placeholder_as_batch_major()
+    start_data = start.output.copy_as_batch_major()  # e.g. (B,) or (B,T)
+    start_t = start_data.placeholder
     if size is None:
       if seq_lens is None:
-        size = tf.maximum(tf.reduce_max(x.batch_shape[1] - start), 0)
+        size = tf.maximum(tf.reduce_max(x.batch_shape[1] - start_t), 0)  # scalar
       else:
-        size = tf.maximum(tf.reduce_max(seq_lens - start), 0)
+        # make seq_lens compatible with start_t
+        seq_lens = expand_multiple_dims(  # e.g. (B,) or (B,1)
+          x=seq_lens,
+          axes=[-1] * (len(start_t.shape) - len(seq_lens.shape)))
+        size = tf.maximum(tf.reduce_max(seq_lens - start_t), 0)  # scalar
       if min_size is not None:
         size = tf.maximum(size, min_size)
-    self.size = size
-    start = tf.expand_dims(start, axis=1)  # (B, T)
-    slices = slice_nd(x.placeholder, start=tf.cast(start, tf.int32), size=size)  # (B,size, ...)
+    # for each start index in start_data, we want to gather a slice
+    # therefore, the output's first axes are the same as the ones from start_data
+    # and the next axis will therefore be the slice axis
+    slice_tag = self.output.dim_tags[start_data.batch_ndim]
+    assert slice_tag.description.startswith("sliced-time:")
+    if not isinstance(size, int):
+      # in this case, size is not known before runtime and becomes dynamic and we need to set dyn_size
+      if len(seq_lens.shape) == 1:
+        dyn_size = tf.maximum(seq_lens - start_t, 0)  # (B,)
+      else:
+        # in this case, we would normally get a dynamic size of shape (B,T) for the slices
+        # in order to get shape (B,) instead, we reduce all other axes except the batch axis
+        reduce_axes = range(1, len(seq_lens.shape))
+        dyn_size = tf.maximum(tf.reduce_max(seq_lens - start_t, axis=reduce_axes), 0)  # (B,)
```

Comment:
I somehow skipped this before. I don't understand this logic. Why do you reduce the axes? If `seq_lens` was sth more complex before (not just [B]), it should stay that way.

Reply:
Above, I do `seq_lens = expand_multiple_dims(...)` in order to match it to `start_t`. So instead of `if len(seq_lens.shape) == 1:`, I could also do `if len(start_t.shape) == 1:`. So I initially assume that `seq_lens` is of shape `[B]` but this could be wrong, right?

Reply:
But you cannot simply reduce axes to match some shape. Also, `seq_lens` is not necessarily of shape [B].
Basically, it is not of shape [B] exactly here. Just do not do the axes-reduction.

Reply:
So I just tried leaving out the reduction. But then, in the loop-optimized case, `dyn_size` is set to shape `[B,T]` and RETURNN is complaining that `dyn_size.shape.ndims` should be `1`. But that is what you meant when you said that this case is not implemented yet, right?

Reply:
No, what is not implemented is when the dyn size is [B] inside the loop, and this layer gets not moved out, but is being accumulated because it is requested as output layer. This accumulation of dyn seq lengths inside the loop is not implemented.

We do support dyn seq lengths of other shapes (since e.g. #579, an example is #589). At least in principle. Not all layers might handle it correctly.

The error you get is probably because you assign `dyn_size` directly. Instead, you should assign `dyn_size_ext`.

Reply:
What is the correct way of setting `dyn_size_ext`? I tried

```
dyn_size_ext = Data(
        name=("%s:dyn_size" % slice_tag.description),
        dtype=Data.size_dtype,
        placeholder=dyn_size,
        shape=start_data.shape,
        batch_dim_axis=0,
        batch=slice_tag.batch,
        beam=self.output.beam,
        control_flow_ctx=slice_tag.control_flow_ctx)
slice_tag.dyn_size_ext = dyn_size_ext
```

This works for `optimize_move_layers_out = False` but not if the loop is optimized.

Reply:
Instead of using `shape=start_data.shape`, use `dim_tags=start_data.dim_tags`. `batch_dim_axis` is not needed then as well.

It looks correct. But this logic is still a bit experimental. I cannot tell what the problem is with further information. What is the error?

Reply:
It would be good if you prepare the simple test case and push it here to the PR. Then I can also directly see the error.

Reply:
> Instead of using `shape=start_data.shape`, use `dim_tags=start_data.dim_tags`. `batch_dim_axis` is not needed then as well.

This works.

> It looks correct. But this logic is still a bit experimental. I cannot tell what the problem is with further information. What is the error?

I will prepare the test case now and push it to the PR. Should I then also push my `dyn_size_ext` solution so that you can see the error or should I directly push your proposed changes where the error doesn't occur?

Reply:
Push it directly with the proposed changes (with `dyn_size_ext`).